### DISCUSSION TITLE
How to move the start and end points of a Line without creating garbage?

### DISCUSSION THREAD

**Discussion by stevensrmiller** [AvaloniaUI/Avalonia#17087] [Q&A] [answered]
It's easy to construct two Point objects and a Line connecting them:

    Point p0 = new Point(0, 0);
    Point p1 = new Point(100, 100);

    Line line = new Line {StartPoint = p0, EndPoint = p1, };

But if I want to change the coordinates of either end of the Line, I need to supply a new Point:

    line.StartPoint = new Point(200, 300);

That's a problem if I do it often, like in an animation, because the previous StartPoint becomes garbage and the Garbage Collector has to find it and destroy it.

**[EDIT: stevemonaco made it clear that the previous StartPoint does _not_ become garbage. It is simply overwritten.]**

I'd rather just modify the existing Point's coordinates:

    line.StartPoint.X = 200;
    line.StartPoint.Y = 300;

That's not possible, though, because the X and Y properties of a Point are read only.

Is there a way to alter the coordinates of the StartPoint and EndPoint of a Line object without creating any garbage?

**Comment by stevemonaco** [comment]
`Avalonia.Point` is a `struct`, so no garbage should be produced unless I'm missing something.

**Reply by stevensrmiller** [reply]
Oh, hey, I think you're right about that! When structs are members of classes, they take up space on the heap, but when you assign a struct to another struct, it's copied as a value type. Adjusting the values in the points defining the end of a line doesn't require creating any new objects on the heap. Have to mull this one a bit, but it could be I'm asking about a problem that doesn't exist in the first place, couldn't it? Thanks!

**Reply by stevemonaco** [reply, marked as answer]
> When structs are members of classes, they take up space on the heap, but when you assign a struct to another struct, it's copied as a value type.

Not quite. `struct` will always have value-type copy semantics. When a struct is created as a local variable, that typically goes onto the stack. When it's a member of a class, then it's typically stored inline with the rest of the class members. ie. There's no pointer to a separate struct instance on the heap. There's no separate allocation, gc tracking, or collection specific to the `Point` because it's "part" of the class instance, similar as if a class had two `double` members as opposed to one pointer member to a `Point`.

Edit: Only `ref struct` is guaranteed to live on the stack.

**Reply by stevemonaco** [reply]
You can test with the following, though it shouldn't really be relied upon:

```cs
using System.Runtime.InteropServices;

var size = Marshal.SizeOf(typeof(Myclass));
Console.WriteLine(size);

[StructLayout(LayoutKind.Sequential)]
class Myclass
{
    public Point8D Point { get; }
    public Point8D Point2 { get; }
}

public struct Point8D
{
    public int x0;
    public int x1;
    public int x2;
    public int x3;
    public int x4;
    public int x5;
    public int x6;
    public int x7;
}
```

Then change the `Point8D` to a `class`.

**Reply by stevensrmiller** [reply]
> When it's a member of a class, then it's typically stored inline with the rest of the class members. ie. There's no pointer to a separate struct instance on the heap.

No separate reference, so it adds no additional garbage when the object that contains it is eligible for collection. It does take space out  of whatever resource was used to supply the memory needed for the object. As a value type, it would take up space the same as a float or a byte would (like you said, "inline with the rest").

This article reminded me that not every implementation of a language necessarily even has a stack nor, I suppose, even a heap: https://learn.microsoft.com/en-us/archive/blogs/ericlippert/the-truth-about-value-types.

But I believe you are quite right that structs don't need to be garbage collected. (I suppose an implementation could always create them as objects, copying them as value types instead of reference types, which would mean they would need to be garbage collected at some point if nothing held a reference to them anymore, but that appears to be unnecessary and not how dotnet compiles them, based on a few ildasm investigations I just did.)

Hard to get my head wrapped around structs as locals when one still uses the "new" operator to create them. But the compiler can simply treat them differently from how it treats objects.

Thanks for helping me get a grip on that. Marking your reply as the answer.

**Reply by stevemonaco** [reply]
> I suppose an implementation could always create them as objects

This happens when value types get boxed. ie. Assigned to an `object` property. Also historically happened often with collections of value types before generic types were introduced in .NET 2.
